plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/rest/TrinoIcebergRestCatalogFactory.java

```diff
@@ -0,0 +1,100 @@
+import org.apache.iceberg.rest.auth.OAuth2Properties;
+
+import javax.inject.Inject;
+
+import java.net.URI;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class TrinoIcebergRestCatalogFactory
+        implements TrinoCatalogFactory
+{
+    private static final Logger log = Logger.get(TrinoIcebergRestCatalogFactory.class);
+
+    private final CatalogName catalogName;
+    private final String trinoVersion;
+    private final URI serverUri;
+    private final Optional<String> credential;
+    private final Optional<String> token;
+    private final IcebergRestCatalogConfig.Security security;
+    private final boolean isUniqueTableLocation;
+
+    private volatile RESTSessionCatalog icebergCatalog;
+
+    @Inject
+    public TrinoIcebergRestCatalogFactory(
+            CatalogName catalogName,
+            IcebergRestCatalogConfig restConfig,
+            IcebergConfig icebergConfig,
+            NodeVersion nodeVersion)
+    {
+        this.catalogName = requireNonNull(catalogName, "catalogName is null");
+        this.trinoVersion = requireNonNull(nodeVersion, "nodeVersion is null").toString();
+        requireNonNull(restConfig, "restConfig is null");
+        this.serverUri = restConfig.getBaseUri();
+        this.credential = restConfig.getCredential();
+        this.token = restConfig.getToken();
+        this.security = restConfig.getSecurity();
+        requireNonNull(icebergConfig, "icebergConfig is null");
+        this.isUniqueTableLocation = icebergConfig.isUniqueTableLocation();
+    }
+
+    @Override
+    public TrinoCatalog create(ConnectorIdentity identity)
+    {
+        if (icebergCatalog == null) {
+            synchronized (this) {
+                if (icebergCatalog == null) {
+                    ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+                    properties.put(CatalogProperties.URI, serverUri.toString());
+                    properties.put(CatalogProperties.FILE_IO_IMPL, "org.apache.iceberg.io.ResolvingFileIO");
+                    properties.put("trino-version", trinoVersion);
+                    if (security == IcebergRestCatalogConfig.Security.OAUTH2) {
+                        credential.ifPresent(v -> properties.put(OAuth2Properties.CREDENTIAL, v));
+                        token.ifPresent(v -> properties.put(OAuth2Properties.TOKEN, v));
+                    }
+
+                    icebergCatalog = new RESTSessionCatalog();
+                    icebergCatalog.setConf(ConfigurationUtils.getInitialConfiguration());
+                    try {
+                        icebergCatalog.initialize(catalogName.toString(), properties.buildOrThrow());
+                    }
+                    catch (Exception e) {
+                        icebergCatalog = null;
+                        log.error("REST session catalog initialization failed. Reason: %s", e.getMessage());
+                        throw e;
```

Comment:
Log and rethrow is an anti-pattern. Please just throw.
If you think logging is important, document why.

Reply:
This is addressed via the previous comment since we don't require the try/catch and re-throw anymore